Optional feature to show engagement as percent of enrollment

bradenmacdonald opened the pull request:
**Description**: This PR adds a little bit more information to the course engagement summary screen in the dashboard: When enabled, it displays the weekly active student count as a percentage, in addition to the absolute number.

The percentage of active students for a week is approximated as ([# of active students that week] ÷ [# of enrolled students on the last day of the week] × 100).

In order to keep the code change simple for now, the new data is only displayed in the tables and summaries, and is not shown on the chart.

**JIRA**: [OSPR-720](https://openedx.atlassian.net/browse/OSPR-720)

**Partner Info**: 3rd party Open edX instance. Submitting to minimize code drift but no merge requirement.

**Merge deadline**: Not an edX partner, so best effort.

**Screenshots**:
1. With the new feature enabled:
   ![engagement percentages](https://cloud.githubusercontent.com/assets/945577/8891835/c00e0c78-32ed-11e5-826e-d4d1276d2ac0.png)
2. With the new feature disabled:
   ![engagement percentages disabled](https://cloud.githubusercontent.com/assets/945577/8891833/baf9f7c4-32ed-11e5-9955-5a501eea0803.png)


**Configuration**:
The percentage/enrollment data is only computed and shown if the waffle switch `show_stats_vs_enrollment` is active.

**Test Instructions**:
1. Get a running copy of the dashboard.
2. Import at least a few weeks' worth of engagement and enrollment data from an LMS instance such as your own devstack (for instructions on steps 1 and 2 see [here](https://github.com/open-craft/edx-analytics-devstack#running-the-pipeline))
3. Go to {dashboard url}/admin/waffle/switch/ and enable `show_stats_vs_enrollment`.
4. View a course's engagement summary in the dashboard

**Reviewers**: @mtyaka and TBD

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-dashboard&number=335

mtyaka commented:
@bradenmacdonald :+1:  Can you mention the new `show_stats_vs_enrollment` switch in the README?

bradenmacdonald commented:
Ah, good point. Updated.

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-720](https://openedx.atlassian.net/browse/OSPR-720) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

shnayder commented:
Hi @bradenmacdonald. This looks great. A couple of small requests:

- can we have this always on? I don't want to add another feature flag and have to test multiple code paths unless it's really necessary.
- can you make the font size of the percentage the same as the message on the metric card? It's a bit small currently.

bradenmacdonald commented:
@shnayder Changes made :)

shnayder commented:
Thanks Braden! I'll get this in the queue for the team to do a code review.

bradenmacdonald commented:
This branch had a merge conflict so I just rebased it and added a fix to an issue where the "Active %" was not shown when some days' enrollment data was missing.

dsjen commented:
@bradenmacdonald thank you for this PR!  I'm going to start reviewing this soon (finally).

dsjen commented on analytics_dashboard/core/templatetags/dashboard_extras.py:
```diff
@@ -69,11 +69,12 @@ def render(self, context):
 
 @register.inclusion_tag('summary_point.html')
-def summary_point(value, label, subheading=None, tooltip=None):
+def summary_point(value, label, subheading=None, tooltip=None, value_alternate=None):
```
`value_alternate` makes me think that this value will replace the value.  `additional_value` or something like that might be clearer.

dsjen commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,6 +84,34 @@ def _build_summary(self, api_trends):
+        if summary_enrollment:
+            for key in ('any', 'attempted_problem', 'posted_forum', 'played_video'):
+                if summary.get(key):
+                    summary[key + '_alt'] = "{:.0f}%".format(summary[key] * 100.0 / summary_enrollment)
```
We should be able to utilize https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/core/templatetags/dashboard_extras.py#L23 rather than creating the string for display here.  A major benefit of using https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/core/templatetags/dashboard_extras.py#L23 is that the percentage will be translated properly.

bradenmacdonald replied:
`metric_percentage` would be nice to use but I don't think it will integrate easily with the latest request to format this as "X% of current students" (since django template syntax makes interleaving variables in a translated string and then passing the result to another template tag pretty messy).

So what I've got now is:
```python
percent_value = summary[key] * 100.0 / summary_enrollment
# Translators: Leave '{percentage}' in as a placeholder for the percentage value.
percent_str = _("{percentage}% of current students")
summary[key + '_percent_str'] = percent_str.format(percentage="{:.0f}".format(percent_value))
```

Is that ok?

dsjen commented:
Is there CSS that's missing?  The percentages in the summary cards don't look the same when I run this branch locally.

dsjen commented on analytics_dashboard/courses/tests/test_views/test_engagement.py:
```diff
@@ -26,6 +26,10 @@ class CourseEngagementViewTestMixin(PatchMixin, CourseAPIMixin):  # pylint: disa
     def setUp(self):
         super(CourseEngagementViewTestMixin, self).setUp()
         self.toggle_switch('enable_engagement_videos_pages', True)
+        # This view combines the activity API with the engagement API, so we need to mock both.
```
I think this comment means to reference the "enrollment" instead of "engagement" API.  Also, thanks for the comment!

dsjen commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,6 +84,34 @@ def _build_summary(self, api_trends):
+            elif week['enrollment']:
+                week['active_percent'] = num_active / float(week['enrollment'])
+            else:
+                week['active_percent'] = None  # Avoid dive-by-zero but add an entry for this column so it appears
```
"dive-by-zero" should be divide-by-zero rather.  Dive by zero sounds pretty scary!

dsjen commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,6 +84,34 @@ def _build_summary(self, api_trends):
+        most_recent = trends[-1]
+        summary_enrollment = enrollment_by_day.get(most_recent['weekEnding'])
+        if summary_enrollment:
+            for key in ('any', 'attempted_problem', 'posted_forum', 'played_video'):
```
Rather than listing out the keys here, I think https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/courses/presenters/engagement.py#L24 can be used.

dsjen commented on analytics_dashboard/templates/summary_point.html:
```diff
@@ -9,6 +9,9 @@
       <span class="summary-point-note">{{ subheading }}</span>
     {% endif %}
+    {% if value_alternate %}
+      <span class="summary-point-number">{{ value_alternate }}</span>
```
Adding in the same filters from https://github.com/edx/edx-analytics-dashboard/pull/335/files#diff-6285e9663c95d40966250dafff18d375R6 (e.g. `value|intcomma|truncatechars:13` would make this functionality a bit more flexible.

bradenmacdonald replied:
Good idea but `|intcomma|truncatechars:13` won't work now since we're displaying the text "X% of current students"

dsjen commented:
@bradenmacdonald I've taken a pass through the PR and made some suggestions.  Please let me know if you have any questions.

dsjen commented:
@bradenmacdonald, I chatted with @shnayder & @lamagnifica and have a few requests:

* Update column heading
* Update metric card labeling, font size, and color

Annotated screenshot:
![product-updates](https://cloud.githubusercontent.com/assets/5265058/10674070/11a6727c-78c7-11e5-8461-7378ef00bfec.png)

bradenmacdonald commented:
Thanks for the review @dsjen :)

I'll see if we can schedule a follow-up for next week. CC @antoviaque

antoviaque commented:
Thanks for the review, yes : ) Glad to see it!

@bradenmacdonald Yup, you can move the related task to sprint 40, this way we'll see on Monday how we can fit the incomings reviews.

bradenmacdonald commented:
@dsjen I've pushed a commit to address all your comments. 

Here's an updated screenshot:
![screen shot 2015-10-27 at 9 25 01 pm](https://cloud.githubusercontent.com/assets/945577/10779687/7d922016-7cf1-11e5-9452-72dc78d3b758.png)

dsjen commented:
Thank you, @bradenmacdonald!  This looks great.  Please squash the commits and feel free to merge if you can (I'm happy to merge it in if you're not able to).

bradenmacdonald commented:
Thanks @dsjen. Squashed and tests are green. Merging now.
